// awesome-utils-normal/index.js
/*
 * @Author: batman
 * @Date: 2019-12-11 15:44:07
 * @LastEditors: batman
 * @LastEditTime: 2020-08-21 16:31:48
 * @Description: 入口
 */

import { isPhone, isAndroid, isMobile, slasher, wait } from './src/common';
import { stopBubble } from './src/bubble';
import { debounce, throttle } from './src/debounce_throttle';
import { deepCopy, shallowCopy } from './src/clone/index';
import { Decimal } from './src/decimal/decimal';
import { formatPhone, parseTime, formatTime, formatThousandthMoney, formatCardNo, randomGenerate } from './src/format/index';
import { baseIsEqual, isEqual, uniq, uniqWith, uniqBy } from './src/handler/index';
import { Validator } from './src/validation/validator';
import { isIdNo, validatenull } from './src/validation/validate';

export {
	isPhone,
	isAndroid,
	isMobile,
	slasher,
	wait,
	stopBubble,
	debounce,
	throttle,
	deepCopy,
	shallowCopy,
	Decimal,
	formatPhone,
	parseTime,
	formatTime,
	formatThousandthMoney,
	formatCardNo,
	randomGenerate,
	baseIsEqual,
	isEqual,
	uniq,
	uniqWith,
	uniqBy,
	Validator,
	isIdNo,
	validatenull,
};
